ports/rp2/main.c

```diff
@@ -296,3 +337,93 @@ const char rp2_help_text[] =
     "For further help on a specific object, type help(obj)\n"
     "For a list of available modules, type help('modules')\n"
 ;
+
+
+#if MICROPY_HW_USB_VENDOR
+// Invoked when device is mounted
+void tud_mount_cb(void) { }
+
+// Invoked when device is unmounted
+void tud_umount_cb(void) { }
+
+// Invoked when usb bus is suspended
+// remote_wakeup_en : if host allow us  to perform remote wakeup
+// Within 7ms, device must draw an average of current less than 2.5 mA from bus
+void tud_suspend_cb(bool remote_wakeup_en) {
+    (void) remote_wakeup_en;
+}
+
+// Invoked when usb bus is resumed
+void tud_resume_cb(void) { }
+
+// Invoked when a control transfer occurred on an interface of this class
+// Driver response accordingly to the request and the transfer stage (setup/data/ack)
+// return false to stall control endpoint (e.g unsupported request)
+bool tud_vendor_control_xfer_cb(uint8_t rhport, uint8_t stage, tusb_control_request_t const * request) {
+    // Nothing to with DATA & ACK stage
+    if (stage != CONTROL_STAGE_SETUP) return true;
+
+    switch (request->bmRequestType_bit.type)
+    {
+        case TUSB_REQ_TYPE_VENDOR:
+            switch (request->bRequest)
+            {
+                case VENDOR_REQUEST_WEBUSB:
+					// Match vendor request in BOS descriptor
+					// Get landing page url
+					return tud_control_xfer(rhport, request, (void*)(uintptr_t) &desc_url, desc_url.bLength);
+
+                case VENDOR_REQUEST_MICROSOFT:
+					if (request->wIndex == 7) 
+					{
+						// Get Microsoft OS 2.0 compatible descriptor
+						uint16_t total_len;
+						memcpy(&total_len, desc_ms_os_20 + 8, 2);
+
+						return tud_control_xfer(rhport, request, (void*)(uintptr_t) desc_ms_os_20, total_len);
+					} else {
+
+						return false;
+					}
+
+                default: break;
+            }
+        break;
+
+        case TUSB_REQ_TYPE_CLASS:
+			if (request->bRequest == 0x22) {
```

Comment:
Does `0x22` have a human-readable representation?

Reply:
This has been updated to a more readable format.